File: python/logbrew_py/src/logbrew_sdk/_support_ticket.py
"""Local-only support ticket draft helpers for future backend routes."""

from __future__ import annotations

import math
import re
from collections.abc import Callable, Mapping
from typing import Any, Literal, Protocol, TypeAlias, TypedDict, TypeGuard
from urllib.parse import urlsplit

SupportTicketSource: TypeAlias = Literal["cli", "sdk", "website", "docs", "mobile"]
SupportTicketCategory: TypeAlias = Literal[
    "sdk_install_failure",
    "ingest_failure",
    "auth_failure",
    "project_setup",
    "dashboard_issue",
    "docs_confusion",
    "cli_issue",
    "mobile_issue",
    "billing_question",
    "other",
]
SupportDiagnosticsValue: TypeAlias = (
    str | int | float | bool | None | list["SupportDiagnosticsValue"] | dict[str, "SupportDiagnosticsValue"]
)


class SupportTicketDraft(TypedDict, total=False):
    """Planned support-ticket create payload built locally without opening a ticket."""

    project_id: str
    source: SupportTicketSource
    category: SupportTicketCategory
    title: str
    description: str
    environment: str
    runtime: str
    framework: str
    sdk_package: str
    sdk_version: str
    release: str
    trace_id: str
    event_id: str
    diagnostics: dict[str, SupportDiagnosticsValue]


class CreateSupportTicketDraft(Protocol):
    def __call__(
        self,
        *,
        source: SupportTicketSource,
        category: SupportTicketCategory,
        title: str,
        description: str,
        project_id: str | None = None,
        environment: str | None = None,
        runtime: str | None = None,
        framework: str | None = None,
        sdk_package: str | None = None,
        sdk_version: str | None = None,
        release: str | None = None,
        trace_id: str | None = None,
        event_id: str | None = None,
        diagnostics: Mapping[str, Any] | None = None,
    ) -> SupportTicketDraft: ...


SUPPORT_TICKET_SOURCES = {"cli", "sdk", "website", "docs", "mobile"}
SUPPORT_TICKET_CATEGORIES = {
    "sdk_install_failure",
    "ingest_failure",
    "auth_failure",
    "project_setup",
    "dashboard_issue",
    "docs_confusion",
    "cli_issue",
    "mobile_issue",
    "billing_question",
    "other",
}

_MAX_DEPTH = 5
_MAX_STRING_LENGTH = 500
_MAX_SEQUENCE_ITEMS = 20
_REDACTED = "[redacted]"


class _OmitValue:
    __slots__ = ()


_OMIT = _OmitValue()
_SanitizedValue: TypeAlias = SupportDiagnosticsValue | _OmitValue

_SENSITIVE_KEYS = {
    "apikey",
    "auth",
    "authorization",
    "authtoken",
    "bearer",
    "clientsecret",
    "connectionstring",
    "cookie",
    "credential",
    "credentials",
    "dsn",
    "password",
    "passwd",
    "privatekey",
    "refreshtoken",
    "secret",
    "session",
    "setcookie",
    "token",
}
_SENSITIVE_ASSIGNMENT_PATTERN = re.compile(
    r"(?:authorization|api[_-]?key|token|secret|password|passwd|cookie)\s*[:=]",
    re.IGNORECASE,
)
_TOKEN_PATTERN = re.compile(
    r"(?:\bBearer\s+[A-Za-z0-9._~+/=-]+|"
    r"\blbw_ingest_[A-Za-z0-9._-]+|"
    r"\b(?:sk|pk|xox[abprs]?)-[A-Za-z0-9_-]{10,}|"
    r"\bAKIA[0-9A-Z]{16}\b)",
    re.IGNORECASE,
)
_URL_PATTERN = re.compile(r"https?://[^\s\"'<>]+", re.IGNORECASE)
_POSIX_PATH_PATTERN = re.compile(r"(?<![\w.-])(?:/Users|/home|/var/folders|/private/var|/tmp)/[^\s\"'<>]+")
_WINDOWS_PATH_PATTERN = re.compile(r"\b[A-Za-z]:\\[^\s\"'<>]+")


def build_create_support_ticket_draft(
    *,
    sdk_error_type: type[Exception],
    require_allowed_value: Callable[[str, Any, set[str]], None],
    require_non_empty: Callable[[str, Any], None],
    require_trace_id: Callable[[Any], None],
) -> CreateSupportTicketDraft:
    def create_support_ticket_draft(
        *,
        source: SupportTicketSource,
        category: SupportTicketCategory,
        title: str,
        description: str,
        project_id: str | None = None,
        environment: str | None = None,
        runtime: str | None = None,
        framework: str | None = None,
        sdk_package: str | None = None,
        sdk_version: str | None = None,
        release: str | None = None,
        trace_id: str | None = None,
        event_id: str | None = None,
        diagnostics: Mapping[str, Any] | None = None,
    ) -> SupportTicketDraft:
        require_allowed_value("support ticket source", source, SUPPORT_TICKET_SOURCES)
        require_allowed_value("support ticket category", category, SUPPORT_TICKET_CATEGORIES)
        require_non_empty("support ticket title", title)
        require_non_empty("support ticket description", description)

        draft: SupportTicketDraft = {
            "source": source,
            "category": category,
            "title": title.strip(),
            "description": description.strip(),
        }
        project_id = _clean_optional_string("support ticket project_id", project_id, require_non_empty)
        if project_id is not None:
            draft["project_id"] = project_id
        environment = _clean_optional_string("support ticket environment", environment, require_non_empty)
        if environment is not None:
            draft["environment"] = environment
        runtime = _clean_optional_string("support ticket runtime", runtime, require_non_empty)
        if runtime is not None:
            draft["runtime"] = runtime
        framework = _clean_optional_string("support ticket framework", framework, require_non_empty)
        if framework is not None:
            draft["framework"] = framework
        sdk_package = _clean_optional_string("support ticket sdk_package", sdk_package, require_non_empty)
        if sdk_package is not None:
            draft["sdk_package"] = sdk_package
        sdk_version = _clean_optional_string("support ticket sdk_version", sdk_version, require_non_empty)
        if sdk_version is not None:
            draft["sdk_version"] = sdk_version
        release = _clean_optional_string("support ticket release", release, require_non_empty)
        if release is not None:
            draft["release"] = release
        if trace_id is not None:
            require_trace_id(trace_id)
            draft["trace_id"] = trace_id.lower()
        event_id = _clean_optional_string("support ticket event_id", event_id, require_non_empty)
        if event_id is not None:
            draft["event_id"] = event_id
        if diagnostics is not None:
            if not isinstance(diagnostics, Mapping):
                raise sdk_error_type("validation_error", "support ticket diagnostics must be an object")
            draft["diagnostics"] = _sanitize_diagnostics(diagnostics)
        return draft

    create_support_ticket_draft.__name__ = "create_support_ticket_draft"
    create_support_ticket_draft.__doc__ = (
        "Build a local-only, token-free support-ticket create payload draft without calling backend routes."
    )
    return create_support_ticket_draft


def _clean_optional_string(
    label: str,
    value: str | None,
    require_non_empty: Callable[[str, Any], None],
) -> str | None:
    if value is None:
        return None
    require_non_empty(label, value)
    return value.strip()


def _sanitize_diagnostics(diagnostics: Mapping[str, Any]) -> dict[str, SupportDiagnosticsValue]:
    safe: dict[str, SupportDiagnosticsValue] = {}
    for key, value in diagnostics.items():
        if not isinstance(key, str) or not key:
            continue
        if _is_sensitive_key(key):
            safe[key] = _REDACTED
            continue
        sanitized = _sanitize_value(value, 0)
        if _is_support_diagnostics_value(sanitized):
            safe[key] = sanitized
    return safe


def _sanitize_value(value: Any, depth: int) -> _SanitizedValue:
    if depth > _MAX_DEPTH:
        return _OMIT
    sanitized_primitive = _sanitize_primitive(value)
    if sanitized_primitive is not _OMIT:
        return sanitized_primitive
    if isinstance(value, BaseException):
        return {"type": type(value).__name__}
    if isinstance(value, Mapping):
        return _sanitize_mapping(value, depth)
    if isinstance(value, (list, tuple)):
        return _sanitize_sequence(value, depth)
    return _OMIT


def _sanitize_primitive(value: Any) -> _SanitizedValue:
    if isinstance(value, bool) or value is None:
        return value
    if isinstance(value, int):
        return value
    if isinstance(value, float):
        return value if math.isfinite(value) else _OMIT
    if isinstance(value, str):
        return _sanitize_string(value)
    return _OMIT


def _sanitize_mapping(value: Mapping[Any, Any], depth: int) -> dict[str, SupportDiagnosticsValue]:
    safe: dict[str, SupportDiagnosticsValue] = {}
    for key, nested in value.items():
        if not isinstance(key, str) or not key:
            continue
        if _is_sensitive_key(key):
            safe[key] = _REDACTED
            continue
        sanitized = _sanitize_value(nested, depth + 1)
        if _is_support_diagnostics_value(sanitized):
            safe[key] = sanitized
    return safe


def _sanitize_sequence(value: list[Any] | tuple[Any, ...], depth: int) -> list[SupportDiagnosticsValue]:
    safe_items: list[SupportDiagnosticsValue] = []
    for item in value[:_MAX_SEQUENCE_ITEMS]:
        sanitized = _sanitize_value(item, depth + 1)
        if _is_support_diagnostics_value(sanitized):
            safe_items.append(sanitized)
    return safe_items


def _is_support_diagnostics_value(value: _SanitizedValue) -> TypeGuard[SupportDiagnosticsValue]:
    return not isinstance(value, _OmitValue)


def _is_sensitive_key(key: str) -> bool:
    normalized = re.sub(r"[^a-z0-9]", "", key.lower())
    return normalized in _SENSITIVE_KEYS


def _sanitize_string(value: str) -> str:
    if _SENSITIVE_ASSIGNMENT_PATTERN.search(value) or _TOKEN_PATTERN.search(value):
        return _REDACTED
    sanitized = _URL_PATTERN.sub(_redact_url, value)
    sanitized = _POSIX_PATH_PATTERN.sub("[redacted-path]", sanitized)
    sanitized = _WINDOWS_PATH_PATTERN.sub("[redacted-path]", sanitized)
    if len(sanitized) > _MAX_STRING_LENGTH:
        return sanitized[: _MAX_STRING_LENGTH - 3] + "..."
    return sanitized


def _redact_url(match: re.Match[str]) -> str:
    parsed = urlsplit(match.group(0))
    path = parsed.path or ""
    return f"[redacted-url]{path}"

File: go/logbrew/support_ticket.go
package logbrew

import (
	"math"
	"net/url"
	"reflect"
	"regexp"
	"strings"
)

// SupportTicketDraftInput describes an explicit local-only draft for planned
// backend support-ticket routes. It does not open a ticket or send telemetry.
type SupportTicketDraftInput struct {
	ProjectID   string
	Source      string
	Category    string
	Title       string
	Description string
	Environment string
	Runtime     string
	Framework   string
	SDKPackage  string
	SDKVersion  string
	Release     string
	TraceID     string
	EventID     string
	Diagnostics map[string]any
}

// SupportTicketDraft is the planned support-ticket create payload after local
// validation and diagnostics redaction.
type SupportTicketDraft struct {
	ProjectID   string         `json:"project_id,omitempty"`
	Source      string         `json:"source"`
	Category    string         `json:"category"`
	Title       string         `json:"title"`
	Description string         `json:"description"`
	Environment string         `json:"environment,omitempty"`
	Runtime     string         `json:"runtime,omitempty"`
	Framework   string         `json:"framework,omitempty"`
	SDKPackage  string         `json:"sdk_package,omitempty"`
	SDKVersion  string         `json:"sdk_version,omitempty"`
	Release     string         `json:"release,omitempty"`
	TraceID     string         `json:"trace_id,omitempty"`
	EventID     string         `json:"event_id,omitempty"`
	Diagnostics map[string]any `json:"diagnostics,omitempty"`
}

var (
	supportTicketSources    = []string{"cli", "sdk", "website", "docs", "mobile"}
	supportTicketCategories = []string{
		"sdk_install_failure",
		"ingest_failure",
		"auth_failure",
		"project_setup",
		"dashboard_issue",
		"docs_confusion",
		"cli_issue",
		"mobile_issue",
		"billing_question",
		"other",
	}
	supportSensitiveKeys = map[string]struct{}{
		"apikey":           {},
		"auth":             {},
		"authorization":    {},
		"authtoken":        {},
		"bearer":           {},
		"clientsecret":     {},
		"connectionstring": {},
		"cookie":           {},
		"credential":       {},
		"credentials":      {},
		"dsn":              {},
		"password":         {},
		"passwd":           {},
		"privatekey":       {},
		"refreshtoken":     {},
		"secret":           {},
		"session":          {},
		"setcookie":        {},
		"token":            {},
	}
	supportSensitiveKeyMarkers = []string{
		"auth",
		"connectionstring",
		"cookie",
		"credential",
		"dsn",
		"password",
		"passwd",
		"privatekey",
		"secret",
		"session",
		"token",
	}
	supportSensitiveAssignmentPattern = regexp.MustCompile(`(?i)(?:authorization|api[_-]?key|token|secret|password|passwd|cookie)\s*[:=]`)
	supportTokenPattern               = regexp.MustCompile(`(?i)(?:\bBearer\s+[A-Za-z0-9._~+/=-]+|\blbw_ingest_[A-Za-z0-9._-]+|\b(?:sk|pk|xox[abprs]?)-[A-Za-z0-9_-]{10,}|\bAKIA[0-9A-Z]{16}\b)`)
	supportURLPattern                 = regexp.MustCompile(`(?i)https?://[^\s"'<>]+`)
	supportPOSIXPathPattern           = regexp.MustCompile(`(?:^|[^\w.-])(?:/Users|/home|/var/folders|/private/var|/tmp)/[^\s"'<>]+`)
	supportWindowsPathPattern         = regexp.MustCompile(`\b[A-Za-z]:\\[^\s"'<>]+`)
)

const (
	supportDiagnosticsMaxDepth       = 5
	supportDiagnosticsMaxStringBytes = 500
	supportDiagnosticsMaxItems       = 20
	supportRedacted                  = "[redacted]"
)

// CreateSupportTicketDraft builds a local-only, token-free support-ticket
// create payload draft without calling backend support routes.
func CreateSupportTicketDraft(input SupportTicketDraftInput) (SupportTicketDraft, error) {
	if err := requireAllowedValue("support ticket source", input.Source, supportTicketSources); err != nil {
		return SupportTicketDraft{}, err
	}
	if err := requireAllowedValue("support ticket category", input.Category, supportTicketCategories); err != nil {
		return SupportTicketDraft{}, err
	}
	if err := requireNonEmpty("support ticket title", input.Title); err != nil {
		return SupportTicketDraft{}, err
	}
	if err := requireNonEmpty("support ticket description", input.Description); err != nil {
		return SupportTicketDraft{}, err
	}
	draft := SupportTicketDraft{
		Source:      input.Source,
		Category:    input.Category,
		Title:       strings.TrimSpace(input.Title),
		Description: strings.TrimSpace(input.Description),
	}
	if value, err := cleanSupportOptionalString("support ticket project_id", input.ProjectID); err != nil {
		return SupportTicketDraft{}, err
	} else {
		draft.ProjectID = value
	}
	if value, err := cleanSupportOptionalString("support ticket environment", input.Environment); err != nil {
		return SupportTicketDraft{}, err
	} else {
		draft.Environment = value
	}
	if value, err := cleanSupportOptionalString("support ticket runtime", input.Runtime); err != nil {
		return SupportTicketDraft{}, err
	} else {
		draft.Runtime = value
	}
	if value, err := cleanSupportOptionalString("support ticket framework", input.Framework); err != nil {
		return SupportTicketDraft{}, err
	} else {
		draft.Framework = value
	}
	if value, err := cleanSupportOptionalString("support ticket sdk_package", input.SDKPackage); err != nil {
		return SupportTicketDraft{}, err
	} else {
		draft.SDKPackage = value
	}
	if value, err := cleanSupportOptionalString("support ticket sdk_version", input.SDKVersion); err != nil {
		return SupportTicketDraft{}, err
	} else {
		draft.SDKVersion = value
	}
	if value, err := cleanSupportOptionalString("support ticket release", input.Release); err != nil {
		return SupportTicketDraft{}, err
	} else {
		draft.Release = value
	}
	if input.TraceID != "" {
		traceID := strings.ToLower(strings.TrimSpace(input.TraceID))
		if err := requireTraceID(traceID); err != nil {
			return SupportTicketDraft{}, err
		}
		draft.TraceID = traceID
	}
	if value, err := cleanSupportOptionalString("support ticket event_id", input.EventID); err != nil {
		return SupportTicketDraft{}, err
	} else {
		draft.EventID = value
	}
	if input.Diagnostics != nil {
		draft.Diagnostics = sanitizeSupportDiagnostics(input.Diagnostics)
	}
	return draft, nil
}

func cleanSupportOptionalString(label, value string) (string, error) {
	if value == "" {
		return "", nil
	}
	if err := requireNonEmpty(label, value); err != nil {
		return "", err
	}
	return strings.TrimSpace(value), nil
}

func sanitizeSupportDiagnostics(diagnostics map[string]any) map[string]any {
	safe := map[string]any{}
	for key, value := range diagnostics {
		if key == "" {
			continue
		}
		if isSensitiveSupportKey(key) {
			safe[key] = supportRedacted
			continue
		}
		if sanitized, ok := sanitizeSupportDiagnosticValue(value, 0); ok {
			safe[key] = sanitized
		}
	}
	if len(safe) == 0 {
		return nil
	}
	return safe
}

func sanitizeSupportDiagnosticValue(value any, depth int) (any, bool) {
	if depth > supportDiagnosticsMaxDepth {
		return nil, false
	}
	if err, ok := value.(error); ok {
		return map[string]any{"type": supportErrorType(err)}, true
	}
	switch typed := value.(type) {
	case nil, bool, string, int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64:
		return sanitizeSupportPrimitive(typed)
	case float32:
		if math.IsInf(float64(typed), 0) || math.IsNaN(float64(typed)) {
			return nil, false
		}
		return typed, true
	case float64:
		if math.IsInf(typed, 0) || math.IsNaN(typed) {
			return nil, false
		}
		return typed, true
	case map[string]any:
		return sanitizeSupportDiagnosticMap(typed, depth)
	case []any:
		return sanitizeSupportDiagnosticSlice(reflect.ValueOf(typed), depth), true
	default:
		reflected := reflect.ValueOf(value)
		if !reflected.IsValid() {
			return nil, true
		}
		switch reflected.Kind() {
		case reflect.Map:
			if reflected.Type().Key().Kind() == reflect.String {
				return sanitizeSupportDiagnosticReflectMap(reflected, depth), true
			}
		case reflect.Slice, reflect.Array:
			return sanitizeSupportDiagnosticSlice(reflected, depth), true
		}
	}
	return nil, false
}

func sanitizeSupportPrimitive(value any) (any, bool) {
	switch typed := value.(type) {
	case string:
		return sanitizeSupportString(typed), true
	default:
		return value, true
	}
}

func sanitizeSupportDiagnosticMap(value map[string]any, depth int) (map[string]any, bool) {
	safe := map[string]any{}
	for key, nested := range value {
		if key == "" {
			continue
		}
		if isSensitiveSupportKey(key) {
			safe[key] = supportRedacted
			continue
		}
		if sanitized, ok := sanitizeSupportDiagnosticValue(nested, depth+1); ok {
			safe[key] = sanitized
		}
	}
	return safe, true
}

func sanitizeSupportDiagnosticReflectMap(value reflect.Value, depth int) map[string]any {
	safe := map[string]any{}
	iter := value.MapRange()
	for iter.Next() {
		key := iter.Key().String()
		if key == "" {
			continue
		}
		if isSensitiveSupportKey(key) {
			safe[key] = supportRedacted
			continue
		}
		if sanitized, ok := sanitizeSupportDiagnosticValue(iter.Value().Interface(), depth+1); ok {
			safe[key] = sanitized
		}
	}
	return safe
}

func sanitizeSupportDiagnosticSlice(value reflect.Value, depth int) []any {
	limit := value.Len()
	if limit > supportDiagnosticsMaxItems {
		limit = supportDiagnosticsMaxItems
	}
	safe := make([]any, 0, limit)
	for index := 0; index < limit; index++ {
		if sanitized, ok := sanitizeSupportDiagnosticValue(value.Index(index).Interface(), depth+1); ok {
			safe = append(safe, sanitized)
		}
	}
	return safe
}

func supportErrorType(err error) string {
	if err == nil {
		return "<nil>"
	}
	errType := reflect.TypeOf(err)
	if errType.Kind() == reflect.Pointer {
		errType = errType.Elem()
	}
	if errType.PkgPath() != "" {
		return errType.PkgPath() + "." + errType.Name()
	}
	return errType.Name()
}

func isSensitiveSupportKey(key string) bool {
	normalized := normalizeSupportKey(key)
	if _, ok := supportSensitiveKeys[normalized]; ok {
		return true
	}
	for _, marker := range supportSensitiveKeyMarkers {
		if strings.Contains(normalized, marker) {
			return true
		}
	}
	return false
}

func normalizeSupportKey(key string) string {
	var normalized strings.Builder
	for _, char := range strings.ToLower(key) {
		if char >= 'a' && char <= 'z' || char >= '0' && char <= '9' {
			normalized.WriteRune(char)
		}
	}
	return normalized.String()
}

func sanitizeSupportString(value string) string {
	if supportSensitiveAssignmentPattern.MatchString(value) || supportTokenPattern.MatchString(value) {
		return supportRedacted
	}
	sanitized := supportURLPattern.ReplaceAllStringFunc(value, redactSupportURL)
	sanitized = supportPOSIXPathPattern.ReplaceAllStringFunc(sanitized, redactSupportPathMatch)
	sanitized = supportWindowsPathPattern.ReplaceAllString(sanitized, "[redacted-path]")
	if len(sanitized) > supportDiagnosticsMaxStringBytes {
		return sanitized[:supportDiagnosticsMaxStringBytes-3] + "..."
	}
	return sanitized
}

func redactSupportURL(value string) string {
	parsed, err := url.Parse(value)
	if err != nil {
		return "[redacted-url]"
	}
	return "[redacted-url]" + parsed.Path
}

func redactSupportPathMatch(value string) string {
	if strings.HasPrefix(value, "/") {
		return "[redacted-path]"
	}
	return value[:1] + "[redacted-path]"
}
